src/protocol/malicious.rs

```diff
@@ -0,0 +1,306 @@
+
+        let r_share = prss.generate_replicated(RecordId::from(0));
+        let (u_left, u_right): (F, F) = prss.generate_fields(RecordId::from(1));
+        let (w_left, w_right): (F, F) = prss.generate_fields(RecordId::from(2));
+
+        SecurityValidator {
+            internally_indexed_ctx,
+            r_share,
+            u: u_right - u_left,
+            w: w_right - w_left,
+        }
+    }
+
+    #[allow(dead_code)]
+    fn compute_dot_product_contribution(a: Replicated<F>, b: Replicated<F>) -> F {
+        let (a_left, a_right) = a.as_tuple();
+        let (b_left, b_right) = b.as_tuple();
+        (a_left + a_right) * (b_left + b_right) - a_right * b_right
+    }
+
+    #[allow(dead_code)]
+    fn accumulate_macs(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+        record_id: RecordId,
+        x: Replicated<F>,
+        rx: Replicated<F>,
+    ) {
+        // The helpers need to use the same shared randomness to generate the random constant used to validate a given multiplication.
+        // This is a bit tricky, because we cannot count on the multiplications being executed in the same order across all the helpers.
+        // The easiest way is to just narrow the context used to perform the multiplication, and then re-use the same record_id.
+        // This ensures that when the helpers all go to validate the multiplication: "1/foo/bar/baz", they all use the prss from "1/foo/bar/baz/validate".
+        // That way, we don't need to worry about the order in which the multiplications are executed.
+        let prss = ctx.narrow(&"validate".to_string()).prss();
+        let random_constant = prss.generate_replicated(record_id);
+
+        self.u += Self::compute_dot_product_contribution(random_constant, rx);
+        self.w += Self::compute_dot_product_contribution(random_constant, x);
+    }
+
+    ///
+    /// For each input, `x` of the arithmetic circuit, call this method to obtain
+    /// a secret-sharing of `r * x`.
+    /// This will also affect the information theoretic MAC that is eventually validated
+    /// once the circuit is complete and before outputs are revealed
+    ///
+    #[allow(dead_code)]
+    pub async fn input(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+        record_id: RecordId,
+        x: Replicated<F>,
+    ) -> Result<Replicated<F>, BoxError> {
+        let rx = ctx
+            .narrow(&"validate_input".to_string())
+            .multiply(record_id)
+            .await
+            .execute(self.r_share, x)
+            .await?;
+
+        self.accumulate_macs(ctx, record_id, x, rx);
+        Ok(rx)
+    }
+
+    ///
+    /// Each time a multiplication is performed, it must be performed twice, in order to compute both the desired result: `z`,
+    /// as well as the result `r * z`. Once these two values are obtained, they should be passed to this function.
+    #[allow(dead_code)]
+    pub fn validate_multiplication_output(
```

Comment:
The reason I made the different was to try to hide `r` from everything else aside from the `SecurityValidator`. This interface allows you to get the _dual_ for all your inputs. Everything else - you just perform operations with those _duals_ to keep getting _dual_ versions of all the intermediate stuff.

If you want to make the interfaces the same, that would seem to require exposing `r`. We can do that, but we just need to be really careful about what people do with it (i.e. they better not `reveal` it).

Reply:
OK, I went with your suggestion to just expose a single function in the public interface, and not treat inputs in a special way. This led to a huge simplification of the code since the SecurityValidator no longer needs to be shared across threads - only the share of "r" needs to be available across threads, and it can be safely cloned.

If we still need to access the SecurityValidator across threads, we can use Arc<Mutex<>>.